### ISSUE TITLE
Proposal: Refactor Resource containers into "hyperresource" container.

### ISSUE THREAD

**Issue by dlorenc** [tektoncd/pipeline#1167]
Currently Resources built into Tekton take the form of a set of containers and volumes that are run in order before and/or after a TaskRun executes. Each Resource is responsible for mutating the TaskRun/Pod spec to add its own init and/or regular containers.

This works, but has led to a sprawl in the number of containers executed, which complicates the generated pod yamls.

This issue proposes refactoring these resource containers into one "mega-container". This container would use an image that has all the dependencies of each resource (tools like gsutil, cp, etc.).

Instead of each resource adding it's own entrypoint and environment variables, each resource should mutate a configuration object which is passed to this mega container. This container has an entrypoint that can execute the resources in the proper order/environment. Some benefits of this:

- An easier to understand Pod spec
- Potentially easier testing. Unit tests can verify a simple, static configuration instead of a complex pod spec
- Performance/error handling. This mega container could potentially run resource steps in parallel or retry basic failures
- Less bash! Right now most resources take the form of bash scripts generated by Go. We could instead run them in Go where possible.

cc @afrittoli

**Comment by afrittoli** [comment]
> Potentially easier testing. Unit tests can verify a simple, static configuration instead of a complex pod spec

\o/ That would be great

**Comment by vdemeester** [comment]
@dlorenc what is the status of this one ? (in relation to the PipelineResource design work happening those days)

**Comment by vdemeester** [comment]
@dlorenc @sbwsg @bobcatfish what is the status for this one ? 
Seems to be stale *and* given the current status of `PipelineResource`, maybe we could close this one and take those requirement / idea into account in the potential re-design of PipelineResource

**Comment by sbwsg** [comment]
agreed, this is now no longer relevant given current trajectory of pipeline resources vis-a-vis the equivalent catalog entries.

**Comment by afrittoli** [comment]
I think it's OK to close this for now. 
Personally I still hope pipeline resources will survive, I feel catalog tasks are not a full replacement for pipeline resources, as they lack the abstraction provided by resources and also they are tasks, which makes them impossible to use with TaskRuns.
